core/store/src/parallel_iter.rs

```diff
@@ -0,0 +1,692 @@
+        assert_eq!(ranges.len(), 3);
+        assert_eq!(ranges[0].to_string(), "00010203..01");
+        assert_eq!(ranges[1].to_string(), "01..02");
+        assert_eq!(ranges[2].to_string(), "02..02030405");
+
+        let ranges = IterationRange::from_str("0100010203..0102030405").unwrap().divide();
+        assert_eq!(ranges.len(), 3);
+        assert_eq!(ranges[0].to_string(), "0100010203..0101");
+        assert_eq!(ranges[1].to_string(), "0101..0102");
+        assert_eq!(ranges[2].to_string(), "0102..0102030405");
+
+        let ranges = IterationRange::from_str("01000000000000000030..0101").unwrap().divide();
+        assert_eq!(ranges.len(), 256);
+        assert_eq!(ranges[0].to_string(), "01000000000000000030..010001");
+        assert_eq!(ranges[1].to_string(), "010001..010002");
+        assert_eq!(ranges[255].to_string(), "0100ff..0101");
+
+        let ranges = IterationRange::from_str("..00").unwrap().divide();
+        assert_eq!(ranges.len(), 1);
+        assert_eq!(ranges[0].to_string(), "..00");
+    }
+
+    #[test]
+    fn test_work_item_divide() {
+        let range = WorkItem::Range(IterationRange::from_str("00..01").unwrap());
+        let items = range.divide();
+        assert_eq!(items.len(), 257);
+        assert_eq!(items[0].to_string(), "range 00..0000");
+        assert_eq!(items[5].to_string(), "range 0004..0005");
+        assert_eq!(items[256].to_string(), "range 00ff..01");
+
+        let item = WorkItem::Lookup(Arc::new(vec![vec![0], vec![1], vec![2]]), 0..3).divide();
+        assert_eq!(item.len(), 3);
+        assert_eq!(item[0].to_string(), "keys 0..1");
+        assert_eq!(item[1].to_string(), "keys 1..2");
+        assert_eq!(item[2].to_string(), "keys 2..3");
+
+        let item = WorkItem::Lookup(Arc::new(vec![]), 0..16).divide();
+        assert_eq!(item.len(), 16);
+        assert_eq!(item[0].to_string(), "keys 0..1");
+        assert_eq!(item[1].to_string(), "keys 1..2");
+        assert_eq!(item[2].to_string(), "keys 2..3");
+        assert_eq!(item[15].to_string(), "keys 15..16");
+
+        let item = WorkItem::Lookup(Arc::new(vec![]), 0..1).divide();
+        assert_eq!(item.len(), 1);
+        assert_eq!(item[0].to_string(), "keys 0..1");
+
+        let item = WorkItem::Lookup(Arc::new(vec![]), 0..17).divide();
+        assert_eq!(item.len(), 16);
+        assert_eq!(item[0].to_string(), "keys 0..1");
+        assert_eq!(item[15].to_string(), "keys 15..17");
+
+        let item = WorkItem::Lookup(Arc::new(vec![]), 0..256).divide();
+        assert_eq!(item.len(), 16);
+        assert_eq!(item[0].to_string(), "keys 0..16");
+        assert_eq!(item[15].to_string(), "keys 240..256");
+    }
+
+    #[test]
+    fn test_parallel_iteration() {
+        let store = create_test_store();
+        let mut update = store.store_update();
+        let mut data = BTreeMap::new();
+        for _ in 0..1000000 {
+            let key_len = rand::random::<usize>() % 10;
```

Comment:
could consider using a hardcoded seed in these rand calls so if it ever fails it will be easy to reproduce